src/celeritas/optical/detail/PreGenAction.cc

```diff
@@ -0,0 +1,166 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2024 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file celeritas/optical/detail/PreGenAction.cc
+//---------------------------------------------------------------------------//
+#include "PreGenAction.hh"
+
+#include <algorithm>
+
+#include "corecel/Assert.hh"
+#include "celeritas/global/ActionLauncher.hh"
+#include "celeritas/global/CoreParams.hh"
+#include "celeritas/global/CoreState.hh"
+#include "celeritas/global/CoreTrackData.hh"
+#include "celeritas/global/TrackExecutor.hh"
+#include "celeritas/optical/CerenkovParams.hh"
+#include "celeritas/optical/OpticalPropertyParams.hh"
+#include "celeritas/optical/ScintillationParams.hh"
+
+#include "OpticalGenStorage.hh"
+#include "PreGenExecutor.hh"
+
+namespace celeritas
+{
+namespace detail
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Construct with action ID, optical properties, and storage.
+ */
+PreGenAction::PreGenAction(ActionId id,
+                           SPConstProperties properties,
+                           SPConstCerenkov cerenkov,
+                           SPConstScintillation scintillation,
+                           SPGenStorage storage)
+    : id_(id)
+    , properties_(std::move(properties))
+    , cerenkov_(std::move(cerenkov))
+    , scintillation_(std::move(scintillation))
+    , storage_(std::move(storage))
+{
+    CELER_EXPECT(id_);
+    CELER_EXPECT(scintillation_ || (cerenkov_ && properties_));
+    CELER_EXPECT(!cerenkov == !properties);
```

Comment:
Not sure whether this check is correct as optical properties (refraction_index?) may be still needed for optical interaction codes (for example, Rayleigh and boundary) even when cerenkov is not activated.

Reply:
Good point, you're right, I'll remove this assertion.